libraries/lib-project-file-io/SqliteSampleBlock.cpp

```diff
@@ -289,6 +298,35 @@ SampleBlockPtr SqliteSampleBlockFactory::DoCreateFromXML(
    return sb;
 }
 
+LibSampleTrack::BlockPtr SqliteSampleBlock::Get()
+{
+   assert(mSampleCount > 0);
+
+   // First try to acquire the cache without locking. If that doesn't succeed,
+   // lock and try again. Can't remember how this pattern is called.
```

Comment:
You are thinking of “double checked locking” but that is considered an anti pattern now because of our-of-order execution possibilities.

https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic

I think a proper solution must use these functions instead somewhere, but they are for shared pointers not weak.

Even the check of the weak pointer before the lock is not thread safe.  The standard library requires different shared and weak pointers to maintain the managed objects control block with thread safety, yes.  But races on the pointers themselves can cause undefined behavior.

So really you must put a lock scope around the weak pointer.  You might use the spin lock class that is already in out libraries.

You might also use two, more limited lock scopes by constructing the cache outside of the lock, and only then locking and assigning.

money thread might lose the race to assign non null, in which rare case, it can just release what it allocated.

Reply:
I have looked into the spinlock implementation, and it calls `yield` every other time, so it's not better. I think we need extend it with a policy, that does not `yield` at all. Yielding make the spinlock at least as expensive as the `mutex`, it should be avoided in threads that have latency guarantees.

Reply:
> I think a proper solution must use these functions instead somewhere, but they are for shared pointers not weak.

You shared a link to a solution for the case where multiple threads access the _same shared_ptr object_. Here a new `shared_ptr` object is created each time - in which case "All member functions (including copy constructor and copy assignment) can be called by multiple threads on different instances of shared_ptr without additional synchronization". (https://en.cppreference.com/w/cpp/memory/shared_ptr)

Reply:
> What could be executed out of order here? According to https://en.cppreference.com/w/cpp/memory/weak_ptr/lock, the creation of the shared ptr is executed atomically, and all other variables before the mutex are local.

"Effectively returns expired() ? shared_ptr<T>() : shared_ptr<T>(*this), executed atomically."

Okay, I learned something -- `lock()` on a `std::weak_ptr` is thread safe.  My memory though is that you can't have unsynchronized usage of a `shared_ptr` object.

Reply:
There were more naive implementations of the double checked locking pattern in older days before C++11 that were later understood to be an anti-pattern.

Reply:
Other related and interesting C++ things to know are the thread safety guarantees in the language specification for the first time construction of static variables inside functions, and the related `std::once_flag` defined in `<mutex>`.

Reply:
"Double checked locking".  Use that in the comment.